src/Avalonia.Controls/SelectionModelChangeSet.cs

```diff
@@ -0,0 +1,170 @@
+﻿using System;
+using System.Collections.Generic;
+
+#nullable enable
+
+namespace Avalonia.Controls
+{
+    internal class SelectionModelChangeSet
+    {
+        private List<SelectionNodeOperation> _changes;
+
+        public SelectionModelChangeSet(List<SelectionNodeOperation> changes)
+        {
+            _changes = changes;
+        }
+
+        public SelectionModelSelectionChangedEventArgs CreateEventArgs()
+        {
+            var deselectedIndexCount = 0;
+            var selectedIndexCount = 0;
+            var deselectedItemCount = 0;
+            var selectedItemCount = 0;
+
+            foreach (var change in _changes)
+            {
+                deselectedIndexCount += change.DeselectedCount;
+                selectedIndexCount += change.SelectedCount;
+
+                if (change.Items != null)
+                {
+                    deselectedItemCount += change.DeselectedCount;
+                    selectedItemCount += change.SelectedCount;
+                }
+            }
+
+            var deselectedIndices = new SelectedItems<IndexPath, SelectionNodeOperation>(
+                _changes,
+                deselectedIndexCount,
+                GetDeselectedIndexAt);
+            var selectedIndices = new SelectedItems<IndexPath, SelectionNodeOperation>(
+                _changes,
+                selectedIndexCount,
+                GetSelectedIndexAt);
+            var deselectedItems = new SelectedItems<object?, SelectionNodeOperation>(
+                _changes,
+                deselectedItemCount,
+                GetDeselectedItemAt);
+            var selectedItems = new SelectedItems<object?, SelectionNodeOperation>(
+                _changes,
+                selectedItemCount,
+                GetSelectedItemAt);
+
+            return new SelectionModelSelectionChangedEventArgs(
+                deselectedIndices,
+                selectedIndices,
+                deselectedItems,
+                selectedItems);
+        }
+
+        private IndexPath GetDeselectedIndexAt(
+            List<SelectionNodeOperation> infos,
+            int index)
+        {
+            static int GetCount(SelectionNodeOperation info) => info.DeselectedCount;
+            static List<IndexRange>? GetRanges(SelectionNodeOperation info) => info.DeselectedRanges;
+            return GetIndexAt(infos, index, GetCount, GetRanges);
```

Comment:
Since C# is such a great language you have to write something like ` GetIndexAt(infos, index, i => GetCount(i), i => GetRanges(i))` to avoid allocating new delegates on every call :)

[Sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgZgAIS6BhOgbyrs7o64AcpsANwCGGGAxR0AKjADOGABSEArAB5sAOwwA+OgHMYGMVACUPTu0pc6AX3PcrXWhLoBZBSYfXL164VJIdJoYdADihkwQAK5aHnQAvLrKANz21mlcMvIKcYlhEdGxJiapjpx2ZV5O9ISSrsRx9j6+nP6BwfkYkTGKnnkpGZyD0nKK4V2FGCX2FTZAA===)

Reply:
Ah yeah, this always trips me up - it just feels like the non-lambda syntax should be better!

Reply:
As far as I know, no allocation occurs for local static functions